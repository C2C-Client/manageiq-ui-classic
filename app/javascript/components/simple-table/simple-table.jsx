import React from 'react';
import PropTypes from 'prop-types';

export const Table = ({ children, ...props }) => <table {...props}>{children}</table>;

Table.propTypes = {
  className: PropTypes.string,
};

Table.defaultProps = {
  className: 'table dataTable table-striped table-bordered table-hover',
};

export const TableHead = ({ children, ...props }) => <thead {...props}>{children}</thead>;

export const TableBody = ({ children, ...props }) => <tbody {...props}>{children}</tbody>;

export const TableRow = ({ children, ...props }) => <tr {...props}>{children}</tr>;

export const TableHeading = ({ children, ...props }) => <th {...props}>{children}</th>;

export const TableCell = ({ children, ...props }) => <td {...props}>{children}</td>;

export const DataDrivenTable = ({
  columns,
  rows,
  id,
  transform,
  ...props
}) => {
  const generatedRows = rows.map(transform).map((row, index) => {
    const handleClick = (e) => {
      if (! row.$onClick) {
        return;
      }

      row.$onClick(e);
      e.preventDefault();
      e.stopPropagation();
    };

    return (
      <TableRow key={`row-${index}`} onClick={handleClick}>
        {columns.map(([key, _content, props = {}], index) => <TableCell {...props} key={`${index}-${key}`}>{row[key]}</TableCell>)}
      </TableRow>
    )
  });

  return (
    <Table id={id} {...props}>
      <TableHead>
        <TableRow>
          {columns.map(([key, content, props = {}]) => (
            <TableHeading {...props} key={key}>{content}</TableHeading>
          ))}
        </TableRow>
      </TableHead>
      <TableBody>
        {generatedRows}
      </TableBody>
    </Table>
  );
};

DataDrivenTable.propTypes = {
  columns: PropTypes.array.isRequired,
  rows: PropTypes.arrayOf(PropTypes.object).isRequired,
  id: PropTypes.string,
  transform: PropTypes.func,
};

DataDrivenTable.defaultProps = {
  transform: (row) => row,
};
