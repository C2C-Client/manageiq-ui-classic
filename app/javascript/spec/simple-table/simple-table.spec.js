import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { DataDrivenTable } from '../../components/simple-table/simple-table';

describe('Simple react table', () => {
  let initialProps;
  const clickSpy = jest.fn();
  beforeEach(() => {
    initialProps = {
      columns: [
        ['column-one', 'First label'],
        ['column-two', 'Second label', { className: 'custom-className-of-cells' }],
      ],
      rows: [{
        'column-one': 'Row 1 first column',
        'column-two': <span>Custom component for column 2 row 1</span>,
      }, {
        'column-one': 'Row 2 second column',
        'column-two': <button type="button" onClick={() => clickSpy('foo')} className="button pf-button">Click me!</button>,
      }],
    };
  });

  afterEach(() => {
    clickSpy.mockReset();
  });

  it('should render correctly', () => {
    const wrapper = mount(<DataDrivenTable {...initialProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call callback on custom button cell component', () => {
    const wrapper = mount(<DataDrivenTable {...initialProps} />);
    wrapper.find('button').simulate('click');
    expect(clickSpy).toHaveBeenCalledWith('foo');
  });
});
