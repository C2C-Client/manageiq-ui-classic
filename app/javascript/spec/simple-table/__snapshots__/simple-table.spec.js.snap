// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple react table should render correctly 1`] = `
<Component
  columns={
    Array [
      Array [
        "column-one",
        "First label",
      ],
      Array [
        "column-two",
        "Second label",
        Object {
          "className": "custom-className-of-cells",
        },
      ],
    ]
  }
  rows={
    Array [
      Object {
        "column-one": "Row 1 first column",
        "column-two": <span>
          Custom component for column 2 row 1
        </span>,
      },
      Object {
        "column-one": "Row 2 second column",
        "column-two": <button
          className="button pf-button"
          onClick={[Function]}
          type="button"
        >
          Click me!
        </button>,
      },
    ]
  }
  transform={[Function]}
>
  <Component
    className="table dataTable table-striped table-bordered table-hover"
  >
    <table
      className="table dataTable table-striped table-bordered table-hover"
    >
      <Component>
        <thead>
          <Component>
            <tr>
              <Component
                key="column-one"
              >
                <th>
                  First label
                </th>
              </Component>
              <Component
                className="custom-className-of-cells"
                key="column-two"
              >
                <th
                  className="custom-className-of-cells"
                >
                  Second label
                </th>
              </Component>
            </tr>
          </Component>
        </thead>
      </Component>
      <Component>
        <tbody>
          <Component
            key="row-0"
            onClick={[Function]}
          >
            <tr
              onClick={[Function]}
            >
              <Component
                key="0-column-one"
              >
                <td>
                  Row 1 first column
                </td>
              </Component>
              <Component
                className="custom-className-of-cells"
                key="1-column-two"
              >
                <td
                  className="custom-className-of-cells"
                >
                  <span>
                    Custom component for column 2 row 1
                  </span>
                </td>
              </Component>
            </tr>
          </Component>
          <Component
            key="row-1"
            onClick={[Function]}
          >
            <tr
              onClick={[Function]}
            >
              <Component
                key="0-column-one"
              >
                <td>
                  Row 2 second column
                </td>
              </Component>
              <Component
                className="custom-className-of-cells"
                key="1-column-two"
              >
                <td
                  className="custom-className-of-cells"
                >
                  <button
                    className="button pf-button"
                    onClick={[Function]}
                    type="button"
                  >
                    Click me!
                  </button>
                </td>
              </Component>
            </tr>
          </Component>
        </tbody>
      </Component>
    </table>
  </Component>
</Component>
`;
